core/src/test/scala/integration/kafka/api/GroupAuthorizerIntegrationTest.scala

```diff
@@ -124,12 +128,56 @@ class GroupAuthorizerIntegrationTest extends BaseRequestTest {
       () => producer.send(new ProducerRecord[Array[Byte], Array[Byte]](topic, "message".getBytes)).get()).getCause
     assertTrue(produceException.isInstanceOf[TopicAuthorizationException])
     assertEquals(Set(topic), produceException.asInstanceOf[TopicAuthorizationException].unauthorizedTopics.asScala)
+  }
 
-    val consumer = createConsumer(configsToRemove = List(ConsumerConfig.GROUP_ID_CONFIG))
+  @ParameterizedTest(name = TestInfoUtils.TestWithParameterizedQuorumAndGroupProtocolNames)
+  @MethodSource(Array("getTestQuorumAndGroupProtocolParametersAll"))
+  def testUnauthorizedConsumeUnsubscribe(quorum: String, groupProtocol: String): Unit = {
+    val topic = "topic"
+    val topicPartition = new TopicPartition(topic, 0)
+
+    createTopic(topic, listenerName = interBrokerListenerName)
+
+    val consumer = createConsumer()
+    consumer.assign(List(topicPartition).asJava)
+    val consumeException = assertThrows(classOf[TopicAuthorizationException],
+      () => TestUtils.pollUntilAtLeastNumRecords(consumer, numRecords = 1))
+    assertEquals(Set(topic), consumeException.unauthorizedTopics.asScala)
+
+    assertThrows(classOf[GroupAuthorizationException],
+      () => TestUtils.pollUntilAtLeastNumRecords(consumer, numRecords = 1))
+
+    // TODO: use background-event-queue-size metric to check there is background event
+    Thread.sleep(3000)
+
+    assertDoesNotThrow(new Executable {
+      override def execute(): Unit = consumer.unsubscribe()
+    })
```

Comment:
uhm I think we may be not getting the classic consumer behaviour right here. I would expect this assertion passes for the classic consumer here only because of how the test is written (never finds a coordinator), but that doesn't mean that the classic consumer does not throw topic or group auth exceptions on unsubscribe (which is what the changes in this PR are trying to achieve for the new consumer)

1. I believe classic does not throw GroupAuthException here only because it never finds a coordinator, so the unsubscribes perform no action on the group (but if it had had a known coordinator, the unsubscribe would send a leave group that I expect would fail with GroupAuthException)
2. I believe classic does not throw TopicAuthException here only because unsubscribe does not poll the network client if it doesn't have a coordinator to send the leave group to. 

I could definitely be missing something, but we could validate my expectations with an integration test here:  

- consumer subscribes to a group successfully (has acls to READ + GROUP + "group-name-from-config")
- looses the acls (using `removeAndVerifyAcls`)
- consumer.unsubscribe -> I would expect that this should indeed throw a group authorization exception received in the response to the leave group, or the topic auth exception propagated from metadata (honestly not sure about the precedence, but both should be there)

Reply:
Hi @lianetm, thanks for the review and suggestion. I try to separate it with 4 test cases:

* testConsumeUnsubscribeWithoutTopicPermission
* testConsumeCloseWithoutTopicPermission
* testConsumeUnsubscribeWithoutGroupPermission
* testConsumeCloseWithoutGroupPermission

Although these 4 cases can pass, I need some time to check whether it can really test classic consumer.

Reply:
sharing more details about my expectations of what the classic consumer does (so the contract we should keep):

1. do not throw topic or group auth exceptions on unsubscribe simply because it does not wait for the leave group response after sending the request on `maybeLeaveGroup`
https://github.com/apache/kafka/blob/8cbd2edfe782997380683cbfa7451a4f2de893f0/clients/src/main/java/org/apache/kafka/clients/consumer/internals/ClassicKafkaConsumer.java#L567 
3. could throw topic/group auth on close because it does wait for responses (inflight requests) after the `maybeLeaveGroup`
https://github.com/apache/kafka/blob/8cbd2edfe782997380683cbfa7451a4f2de893f0/clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java#L1140

So we need to throw those on close if they exist, but not on unsubscribe (let's align the expectations on the test with this understanding I would say, and we'll validate it). Let me know what you think/find.

Reply:
Hi @lianetm, thanks for sharing the details in `ClassicKafkaConsumer`. For `ClassicKafkaConsumer#unsubscribe`, it doesn't check future object from `maybeLeaveGroup`, so `unsubscribe` doesn't throw exceptions. I remain `unsubscribe` function test for `AsyncKafkaConsumer` only, and keep `close` function test for both consumers.

I have another question here. IIUC, we don't need to swallow `TopicAuthorizationException`, because we don't check topic permission when doing consumer group heartbeat request. So I think we only need to swallow `GroupAuthorizationException` here.  WDYT? Thanks.

https://github.com/apache/kafka/blob/9db5ed00a8369d5c696e836661230110ea2ea44d/core/src/main/scala/kafka/server/KafkaApis.scala#L3812-L3815